Packages/MIES/MIES_GuiPopupMenuExt.ipf

```diff
@@ -0,0 +1,623 @@
+///        Button popupext_menu3, pos={3.00, 90.00}, size={200,20},proc=PEXT_ButtonProc,title="DropDownMenu3 ▼", userdata=(PEXT_UDATA_ITEMGETTER + ":GetMenuList3;" + PEXT_UDATA_POPUPPROC + ":DemoProc")
+///    End
+///
+///    Function DemoProc(pa) : PopupMenuControl
+///        STRUCT WMPopupAction &pa
+///
+///        print "EventCode/Win/Selected/ctrlName: ", pa.eventCode, pa.win, pa.popStr, pa.ctrlName
+///    End
+///
+///    Function/S GetMenuList1()
+///
+///        Make/FREE/T/N=300 testItems = num2char(97 + mod(p, 26)) + "_" + num2str(p)
+///        return TextWaveToList(testItems, ";")
+///    End
+///
+///    Function/WAVE GetMenuList2()
+///
+///        Make/FREE/T/N=(2) menus
+///        menus[0] = "Kriemhild;Brünhild"
+///        menus[1] = "Siegfried;Gunther"
+///        SetDimLabel ROWS, 0, $"Women", menus
+///        SetDimLabel ROWS, 1, $"Men", menus
+///
+///        return menus
+///    End
+///
+///    Function/WAVE GetMenuList3()
+///
+///        Make/FREE/T menus = {"Kriemhild", "Brünhild", "Siegfried", "Gunther"}
+///        WAVE/T splittedMenu = PEXT_SplitToSubMenus(menus, method = PEXT_SUBSPLIT_ALPHA)
+///        PEXT_GenerateSubMenuNames(splittedMenu)
+///        return splittedMenu
+///    End
+/// \endrst
+///
+/// The original popupmenu has to be replaced by a button with the PEXT_ButtonProc as procedure. The userdata of the
+/// button stores as list with key value pairs a function that returns the menu item list and a function that is the
+/// PopupMenuControl procecure of the former popupmenu. The function for the menu item list corresponds to the function
+/// given for a popupmenu as e.g. value=GetMenuList1() that returns a string with semicolon separated list of menu items.
+/// The names for the submenus are generated automatically in that case.
+/// If one wants to disable the menu, return "" as menu item list. Then a disabled "_none_" is shown that can not be selected.
+///
+///
+/// The constants PEXT_UDATA_ITEMGETTER and PEXT_UDATA_POPUPPROC were defined for the keys in the userdata string.
+///
+/// When a user selects an entry in the context menu the PopupMenuControl control defined through PEXT_UDATA_POPUPPROC is called.
+/// So the proc of the previous popupmenu can be used without further adaptations.
+///
+/// User defined sub menus:
+///
+/// User defined submenus can be created as well. Therefore an extension was added to the function that returns the menu items.
+/// Instead of a string as described before a function can be defined that returns a 1D text wave, e.g. GetMenuList2() in the example.
+/// Each element contains a semicolon separated list of menu items. Each rows DimLabel is used as submenu name, where the elements
+/// menu items are put. The DimLabels must not be empty. The wave can have up to MAX_SUBMENUS elements.
+/// The button popupext_menu2 shows this extension in the example.
```

Comment:
@MichaelHuth @t-b This looks fantastic. Thank you! I would like to implement the user defined lists for the labnotebook but I'm not sure what would happen if GetMenuList2() has list items that don't yet exist in the labnotebook?

Reply:
AddTraceToLBGraph() gets called with the menu entry as key. Therein GetKeyWaveParameterAndUnit() would return NaN and does
`		printf "Could not find %s in keyWave\r", key
`
Then AddTraceToLBGraph() returns NaN and this is silently ignored.

So except the print to the history nothing happens.

Reply:
@MichaelHuth In cases where we know it will be common to have a missing list item, would it be possible to suppress the error message?

Reply:
@timjarsky Probably yes. But as user I would consider present menu items that jsut do nothing inconvenient. 

How about checking for each menu item from the user defined list against the actual present keys. Then we disable menu items that are not present in the LNB by preceding the menu item with a '('. 
See also `DisplayHelpTopic "Special Characters in Menu Item Strings"`

With such approach the menu items would still be visible, just greyed out and not selectable. No change on the handler would be needed.

Reply:
@MichaelHuth Checking and disabling sounds good to me but do you think it would add noticeable lag to the menu?

Reply:
@MichaelHuth @timjarsky I think the current design needs some more explanation before we dive into changing it. The groups for the labnotebook keys are calculated when you click the "Lab Notebook Entries" GUI control. So all the present labnotebook keys can be sorted into their respective group at this point. So from my understanding we don't need to support disabled menu items as we build the groups on the fly from the existing labnotebook keys.